packages/sitecore-jss/src/i18n/graphql.dictionary-service.ts

```diff
@@ -0,0 +1,225 @@
+  ) {
+    total
+    pageInfo {
+      endCursor
+      hasNext
+    }
+    dictionaryPhrases: results {
+      key: field(name: "key") {
+        value
+      },
+      phrase: field(name: "phrase") {
+        value
+      }
+    }
+  }
+}
+`;
+
+export interface GraphQLDictionaryServiceConfig extends CacheOptions {
+  endpoint: string;
+  rootItemId?: string;
+  pageSize?: number;
+  appName: string;
+}
+
+/**
+ * A reply from the GraphQL Sitecore Dictionary Service
+ */
+type DictionaryQueryResult = {
+  search: {
+    pageInfo: {
+      endCursor: string;
+      hasNext: boolean;
+    };
+    dictionaryPhrases: {
+      key: {
+        value: string;
+      };
+      phrase: {
+        value: string;
+      };
+    }[];
+  };
+};
+
+/**
+ * Fetch dictionary data using the Sitecore Dictionary Service GraphQL API.
+ * Uses graphql-request as the default data fetcher (@see GraphQLRequestClient).
+ */
+export class GraphQLDictionaryService extends DictionaryServiceBase {
+  /**
+   * Creates an instance of graphQL dictionary service with the provided options
+   * @param {GraphQLDictionaryService} options instance
+   */
+  constructor(public options: GraphQLDictionaryServiceConfig) {
+    super(options);
+    this.options.pageSize = this.options.pageSize ?? 10;
+  }
+
+  /**
+   * Fetches dictionary data for internalization.
+   * @param {string} language the language to fetch
+   * @default Search query
+   * query DictionarySearch(
+   * $rootItemId: String!,
+   * $language: String!,
+   * $dictionaryEntryTemplateId: String!,
+   * $pageSize: Int = 10,
+   * $after: String
+   * ) {
+   * search(
+   * where: {
+   * AND:[
+   * { name: "_path",      value: $rootItemId },
+   * { name: "_templates", value: $dictionaryEntryTemplateId },
+   * { name: "_language",  value: $language }
+   * ]
+   * }
+   * first: $pageSize
+   * after: $after
+   * orderBy: { name: "Title", direction: ASC }
+   * ) {
+   * total
+   * pageInfo {
+   * endCursor
+   * hasNext
+   * }
+   * dictionaryPhrases: results {
+   * key: field(name: "key") {
+   * value
+   * },
+   * phrase: field(name: "phrase") {
+   * value
+   * }
+   * }
+   * }
+   * }
+   */
+  async fetchDictionaryData(language: string): Promise<DictionaryPhrases> {
+    const cacheKey = this.options.appName + language;
+    const cachedValue = this.getCachedValue(cacheKey);
+    if (cachedValue) {
+      return cachedValue;
+    }
+
+    const dataFetcher = new GraphQLRequestClient(this.options.endpoint);
+    if (!this.options.rootItemId) {
+      this.options.rootItemId = await getAppRoot(dataFetcher, this.options.appName, language);
+    }
+
+    const results = await this.getDictionaryPhrases(dataFetcher, language);
+    return this.cacheValue(cacheKey, results);
+  }
+
+  /**
+   * Gets dictionary phrases
+   * @param {GraphQLRequestClient} dataFetcher
+   * @param {string} language
+   * @returns dictionary phrases
+   */
+  async getDictionaryPhrases(
```

Comment:
I raised this issue in the last sprint demo. Here is why I made the choice not to allow overriding the query:

Originally, our original (rest) dictionary service was doing more than just doing a data fetch. Specifically, it was responsible for caching, and it was responsible for generating the request URL. After the refactoring, responsibilities common to all dictionary services have been moved to the abstract DictionaryService class. And since the GraphQLDictionaryService does't need to generate the endpoint, there is nothing else in GraphQLDictionaryService  other than executing the specific gql query.

So, to do a custom query, it's better for developers to extend DictionaryService with their _custom_ service. If someone has a custom query, there is nothing in GraphQLDictionaryService that would be needed.

Reply:
Thanks for the explanation, I think it's clear this would be overkill here. Unless @illiakovalenko feels strongly otherwise, I'm good considering this resolved.

Reply:
@anastasiya29 Agree, thank you. I resolve it